### DISCUSSION TITLE
cli lib with jbang?

### DISCUSSION THREAD

**Discussion by jmini** [jbangdev/jbang#284] [General]
After having written a couple of jbang scripts, I start to feel the need to model and parse cli inputs correctly (something better than working with the `String[]` argument directly. With named parameters and validation of mandatory parameters)

What is the recommended way for doing this correctly?
Integrate the picocli library? Is there an example somewhere?

**Comment by jmini** [comment]
I have seen [this twitt](https://twitter.com/maxandersen/status/1302010866615943175) from @maxandersen :

> another quarkus *jbang* experiment ... the "minimal" command mode with picocli ... https://gist.github.com/maxandersen/e5a83ec1f19c58f38be5c6d127515bff

Why is quarkus needed? Can't I have a jbang script that is using picocli only?

**Reply by siancu** [reply]
Sure you can. If you execute the following command:

`jbang init --template=cli myCommand.java`

jbang generates a script which is using only picocli without quarkus.

The tweet you refer to is an example of a Quarkus CLI app. 

You can find examples in the examples folder of this repo, for example gh_fetch_release_assets.java.

**Reply by maxandersen** [reply]
yes, @siancu is right on the money. I personally recommend picocli; its what we use in jbang itself and yes we even have a template for it (currently the only extra one).

There are others out there like apache commons cli, and they all should work just fine with jbang - but I just prefer picocli for its simplicity, documentation and all-round awesomeness :)

**Reply by maxandersen** [reply]
oh and this twitter thread from last month might be useful too: https://threadreaderapp.com/thread/1290143961516924933.html 

talks about other things beyond cli parsing that is useful dealing with command line tools in java.
